internal/wavefront/metric/connection.go

```diff
@@ -64,35 +66,62 @@ func (c *Connection) Connect(addr string) error {
 	return nil
 }
 
+func normalizeAddr(addr string) string {
+	if strings.HasPrefix(addr, "http://") || strings.HasPrefix(addr, "https://") {
+		return addr
+	}
+	return "http://" + addr
+}
+
 func (c *Connection) Send(metrics []Metric) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
 	for _, metric := range metrics {
 		c.metrics[metric.Name] = metric
 	}
 }
 
-func (c *Connection) FlushMetrics() {
-	if c.sender == nil {
+func (c *Connection) Flush() {
+	sendBatch(c.extractBatch())
+}
+
+func sendBatch(sender Sender, metrics map[string]Metric) {
+	if sender == nil {
 		return
 	}
-
-	for _, metric := range c.metrics {
-		err := c.sender.SendMetric(metric.Name, metric.Value, 0, metric.Source, metric.Tags)
+	for _, metric := range metrics {
+		err := sender.SendMetric(metric.Name, metric.Value, 0, metric.Source, metric.Tags)
 		if err != nil {
 			log.Log.Error(err, "error sending metrics")
 		}
 	}
-
-	err := c.sender.Flush()
+	err := sender.Flush()
 	if err != nil {
 		log.Log.Error(err, "error flushing metrics")
 	}
 }
 
+func (c *Connection) extractBatch() (Sender, map[string]Metric) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	dst := make(map[string]Metric, len(c.metrics))
+	for key, metric := range c.metrics {
+		dst[key] = metric
+	}
+	return c.sender, dst
+}
+
 func (c *Connection) Close() {
-	if !c.connected() {
-		return
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.close()
+}
+
+// close is not thread safe and must only be called when already holding c.mu
+func (c *Connection) close() {
+	if c.connected() {
+		c.sender.Close()
 	}
-	c.sender.Close()
 	c.sender = nil
 	c.addr = ""
```

Comment:
So it's okay to run these outside of the c.connected() check?

Reply:
sender = nil and addr = "" that is

Reply:
I think so. If sender or addr are already empty, then making them empty again should be benign.

Reply:
Your comment made me realize that I didn't need the `if c.connected()` check in the `Connect` method. I went ahead and removed it.